### PR TITLE
feat: ACME flow skeleton

### LINKED ISSUE #167: Create BULLMQ queue structure (with skeletons) to create a certificate
We need four (nested) steps

In *reverse order, from child to parent*:
- Create order (add DNS records)
- Wait for DNS propagation
- Confirm challenges with the ACME provider
- Finish the order, clean up DNS

Resources:
https://docs.bullmq.io/guide/flows
https://github.com/Seneca-CDOT/starchart/pull/161

### PR DESCRIPTION
**Pull request by dadolhay**
Closes #167

### CODE REVIEW

File: app/queues/notifications.server.ts [hunk 1]
```diff
@@ -44,7 +42,7 @@ notificationsQueue.on('error', (err) => {
  */
 export const notificationsWorker = new Worker<NotificationData>(
   'notifications',
-  async (job: Job) => {
```
Comment (dadolhay) [inline app/queues/notifications.server.ts:47]: type assignment here overrode the `NotificationData` generic passed in when instantiating Worker and reverted job.data to have an explicit `any` type

**Comment by dadolhay** [comment]
@humphd Could you please take a look at the e2e test suite? It passes for me locally, but runs into a weird connection refused issue on CI

File: app/queues/certificate/certificateFlow.server.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+import type { DnsWaiterData } from './dnsWaiterWorker.server';
+import {
+  challengeCompleterQueueName,
+  challengeCompleterWorker,
+} from './challengeCompleterWorker.server';
+import type { ChallengeCompleterData } from './challengeCompleterWorker.server';
+import { orderCompleterQueueName, orderCompleterWorker } from './orderCompleterWorker.server copy';
+import type { OrderCompleterData } from './orderCompleterWorker.server copy';
+import { dnsCleanerQueueName, dnsCleanerWorker } from './dnsCleanerWorker.server';
+import type { DnsCleanerData } from './dnsCleanerWorker.server';
```
Comment (humphd) [inline app/queues/certificate/certificateFlow.server.ts:15]: nit: separate your module imports from type imports vertically in the file to make it easier to read what's happening:

```ts
import { FlowProducer } from 'bullmq';
...

import type { FlowJob } from 'bullmq';
...
```

**Review by humphd — CHANGES_REQUESTED** [review]
Question: does the `FlowProducer` not require you to work with existing queues?  That is, do you not have to add all these queues you reference?  Or does it somehow provide them internally?  My reading of the docs leads me to believe that we need them, but maybe I'm misunderstanding.  If we do need them, should we add them in this PR?

A few small things I noticed reading the code.

Another preference I have is to avoid uppercase in filenames, and prefer using kebab-case, due to being burned on filesystems that aren't case sensitive in the past.

I like how you're approaching this work: splitting it up into stages, creating a skeleton similar to what @Eakam1007 did in the front-end.  That style is fantastic for allowing us to move faster.  Well done.

**Comment by humphd** [comment]
> @humphd Could you please take a look at the e2e test suite? It passes for me locally, but runs into a weird connection refused issue on CI

I'll file an issue, looks like we need to add redis https://github.com/DevelopingSpace/starchart/blob/main/.github/workflows/ci.yaml#L121-L130

File: app/queues/certificate/certificateFlow.server.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  };
+
+  // Step 4, complete order
+  const orderCompleter: FlowJob = {
+    name: `completeOrder:${rootDomain}`,
+    queueName: orderCompleterQueueName,
+    data: { rootDomain } as OrderCompleterData,
+    children: [challengeCompleter],
+    opts: {
+      failParentOnFailure: false, // Important, don't wail the cleanup step
```
[the remark this answers is not in the document]
Comment (dadolhay) [reply app/queues/certificate/certificateFlow.server.ts:93]: Yep, fixing

File: app/queues/certificate/orderCompleterWorker.server copy.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { Worker } from 'bullmq';
```
[the remark this answers is not in the document]
Comment (dadolhay) [reply app/queues/certificate/orderCompleterWorker.server copy.ts:1]: No, copy-paste error, thanks for catching it!

File: app/queues/certificate/certificateFlow.server.ts (hunk 2 above)
Comment (dadolhay) [reply app/queues/certificate/certificateFlow.server.ts:15]: Sure!

**Comment by humphd** [comment]
@Genne23v I'm going to add you to this review as well, since I think you'd find this code interesting.  What @dadolhay is doing here is pretty similar to what you need to do with the DNS flow:

1. create an initial record in MySQL for a domain with pending status
2. talk to Route53 to create/update/delete a record
4. poll Route53 to check connection status of the domain until it's ready
5. update the MySQL record with the active or error status

**Comment by dadolhay** [comment]
> Question: does the `FlowProducer` not require you to work with existing queues? That is, do you not have to add all these queues you reference? Or does it somehow provide them internally? My reading of the docs leads me to believe that we need them, but maybe I'm misunderstanding. If we do need them, should we add them in this PR?
> 
> A few small things I noticed reading the code.
> 
> Another preference I have is to avoid uppercase in filenames, and prefer using kebab-case, due to being burned on filesystems that aren't case sensitive in the past.
> 
> I like how you're approaching this work: splitting it up into stages, creating a skeleton similar to what @Eakam1007 did in the front-end. That style is fantastic for allowing us to move faster. Well done.

In the [docs](https://docs.bullmq.io/guide/flows) there is a section like this:

```js
import { FlowProducer } from 'bullmq';

// A FlowProducer constructor takes an optional "connection"
// object otherwise it connects to a local redis instance.
const flowProducer = new FlowProducer();

const flow = await flowProducer.add({
  name: 'renovate-interior',
  queueName: 'renovate',
  children: [
    { name: 'paint', data: { place: 'ceiling' }, queueName: 'steps' },
    { name: 'paint', data: { place: 'walls' }, queueName: 'steps' },
    { name: 'fix', data: { place: 'floor' }, queueName: 'steps' },
  ],
});
```
> The above code will add atomically 4 jobs, one to the "renovate" queue and 3 to the "steps" queue. When the 3 jobs in the "steps" queue are completed, the parent job in the "renovate" queue will be processed as a regular job.

Based on this, and the fact that the `FlowJob` type is basically an extension of `JobsOptions`, I suspect that this does actually create the queues for us.

**Comment by humphd** [comment]
> Based on this, and the fact that the `FlowJob` type is basically an extension of `JobsOptions`, I suspect that this does actually create the queues for us.

Look at https://github.com/taskforcesh/bullmq/blob/b741e8456f262b51aa7c68f571c76a3c54d02d37/tests/test_flow.ts, it seems to always create the queues it references.

I'm really not sure.

**Comment by humphd** [comment]
Rebase on `main` to pick up the types fix for e2e tests in #208.  Thanks to @shawnyu5 for the quick fix!

**Comment by dadolhay** [comment]
hm, the same issue persists.

**Comment by dadolhay** [comment]
Ok, I see Redis was not yet added

Blocked by #215

**Comment by humphd** [comment]
Apologies, I got two PRs mixed up.

**Comment by dadolhay** [comment]
> Apologies, I got two PRs mixed up.

Nothing to apologize for, happens to everyone!

**Comment by dadolhay** [comment]
Aaand it works!

**Review by humphd — DISMISSED** [review]
I added some code to let me test this locally (feel free to steal this and add to your PR if you want to make manual testing easier):

```diff
diff --git a/app/routes/dev.tsx b/app/routes/dev.tsx
index fda0eaa..f127f22 100644
--- a/app/routes/dev.tsx
+++ b/app/routes/dev.tsx
@@ -13,6 +13,7 @@ import { requireUser, requireUsername } from '~/session.server';
 import { addNotification } from '~/queues/notifications.server';
 
 import type { LoaderArgs, ActionArgs } from '@remix-run/node';
+import { addCertRequest } from '~/queues/certificate/certificate-flow.server';
 
 export const action = async ({ request }: ActionArgs) => {
   const user = await requireUser(request);
@@ -23,6 +24,12 @@ export const action = async ({ request }: ActionArgs) => {
   // If you want do do another action, set a different `intent`
   // in another form below and post.
   switch (intent) {
+    case 'certificate-request':
+      await addCertRequest('testing.starchart.com');
+      return json({
+        result: 'ok',
+        message: 'Certificate requested',
+      });
     case 'send-notification':
       await addNotification({
         emailAddress: user.email,
@@ -60,6 +67,11 @@ export default function Index() {
           <Button type="submit">Send Notification</Button>
         </Form>
 
+        <Form method="post">
+          <input type="hidden" name="intent" value="certificate-request" />
+          <Button type="submit">Request Certificate</Button>
+        </Form>
+
         {data && (
           <Alert status={data.result === 'ok' ? 'success' : 'error'}>
             <AlertIcon />
```

Screenshot 2023-02-14 at 4 36 40 PM

And on the server logs I see:

```
info: TODO create order for testing.starchart.com
info: TODO wait on DNS for testing.starchart.com
info: TODO create order for testing.starchart.com
info: TODO complete order for testing.starchart.com
info: TODO clean DNS for testing.starchart.com
```

Great job!

**Comment by dadolhay** [comment]
thanks for the snippet, added

**Review by humphd — APPROVED** [review]

**Review by Genne23v — APPROVED** [review]
Thanks for making this skeleton and I think it's a very good strategy to split the heavy work into smaller chunks. I should learn this too. And it is very helpful to write my queue for DNS. Code looks all good to me.

**Pull request merged**